src/coreclr/tools/Common/TypeSystem/Common/MethodDesc.Async.cs

```diff
@@ -0,0 +1,183 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System;
+
+namespace Internal.TypeSystem
+{
+    public sealed partial class MethodSignature
+    {
+        // Flags that extend MethodSignatureFlags.
+        // These are not in metadata, just cached values.
+        // They must not conflict with MethodSignatureFlags.
+        private static class MethodSignatureFlagsExtensions
+        {
+            public const MethodSignatureFlags ReturnsTaskOrValueTask = (MethodSignatureFlags)(1 << 30); // Not a real flag, just a cached value
+            public const MethodSignatureFlags ReturnsTaskOrValueTaskMask = (MethodSignatureFlags)(1 << 31); // Not a real flag, just a cached value
+        }
```

Comment:
This cache might be premature, I wouldn't do it yet.

This has problems. I think we already have a bug that `CreateAsyncSignature` will reuse the cache portion of flags despite updating `ReturnType`.